boskos/metrics/resources.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/test-infra/boskos/common"
+	"k8s.io/test-infra/boskos/ranch"
+)
+
+const (
+	// ResourcesMetricName is the name of the Prometheus metric used to monitor Boskos resources.
+	ResourcesMetricName = "boskos_resourcess"
+	// ResourcesMetricDescription is the description for the Prometheus metric used to monitor Boskos resources.
+	ResourcesMetricDescription = "Number of resources recorded in Boskos by resource type and state."
+)
+
+var (
+	// ResourcesMetricLabels is the list of labels used for the Prometheus metric used to monitor Boskos resources.
+	ResourcesMetricLabels = []string{"type", "state"}
+)
+
+type resourcesCollector struct {
+	boskosResources *prometheus.Desc
+	ranch           *ranch.Ranch
+}
+
+// NewResourcesCollector returns a collector which exports the current counts of
+// Boskos resources, segmented by resource type and state.
+func NewResourcesCollector(ranch *ranch.Ranch) prometheus.Collector {
```

Comment:
What keeps us from using a `prometheus.GaugeVec` rather than building a collector ourselves?

Reply:
I started by using a `GaugeVec`, using a similar approach where we'd export metrics for any resource types present in Boskos. The problem is that if a type were deleted (by completely removing its resources), we'd continue to serve stale metrics for that type until Boskos was restarted (and the gauge was recreated).

The first way I attempted to fix this issue was by deleting any label values that didn't match resource types currently in Boskos. But the Prometheus Go API doesn't have any way to list all known values for a label, which suggested this was probably not the right approach. I considered keeping a list of previously-reported types, but that didn't seem great, either.

I stepped back and considered that having a periodic loop to update metrics in the server was also probably the wrong design. I started work to update the metrics in Ranch, when resources were added, deleted, or updated, but realized there were possible consistency issues there (e.g. if something changed the values in the API server without going through the Boskos API directly).

Ultimately, I noticed that latencies for `/metric` are pretty low, the current metrics server calls this API multiple times for each type, and the implementation of `/metric` already lists all resources and then filters out the ones matching the type. I figured that it shouldn't be that much more expensive to just generate this list when Prometheus metrics are scraped. (Your caching work should make this even less of an issue.)

Also, I think this code ended up being fairly easy to understand and easy to test.